enhancements/cluster-logging/performance-tuning.md

```diff
@@ -0,0 +1,238 @@
+
+New field `ClusterLogForwarder.output[].tuning` with the following sub-fields:
+
+#### delivery: Fast|Safe
+
+`Fast`: Minimize resource use (CPU, disk, network) and maximize throughput in normal operation. Use when:
+- It is acceptable to lose some logs due to restarts, failures or heavy load.
+- The logging system will run close to it's limits, and may sometimes be overloaded.
+- It is important to minimize memory, CPU, disk and network costs.
+
+`Safe`: Minimize log loss due to restarts or failures. Use when:
+- It is important to avoid log loss due to restarts or failures.
+- The logging system is not expected to be pushed close to its limits.
+- Resources (memory, CPU, disk) are sufficient for log throughput to exceed expected peak load.
+
+**Note**: The log collector _cannot_ guarantee reliable end-to-end delivery.
+It has no control over the reliability or throughput of sources and sinks.
+The `delivery` policy tunes the log collector's "best effort".
+
+#### compress: string
+
+Enable compression if supported
+- Value is compression "mode" string, format depends on output type.
+- All outputs accept `compress: true` to enable a default mode.
+- Error if output does not support compression or does not recognize the value.
+
+#### batch: object
+A "batch" is the content of a single "write" or "send" operation.
+Defaults are determined by the underlying sink.
+
+- `maxRecords`: Max number of records to include in a batch.
+- `maxBytes`: Max size in bytes of a batch
+
+#### reconnect: object
+Controls connect and reconnect attempts.
+Defaults are determined by the underlying sink.
+
+- `minDelay`: minimum delay between attempts.
+- `maxDelay`: maximum delay between attempts.
+
+`minDelay` for the first attempt, delay increases up to `maxDelay` and then repeats.
+Back-off algorithm is determined by the underlying sink.
+
+### Implementation Details
+
+#### Causes of log loss
+
+1. *Overload*: Logs are produced faster than they can be processed.
+2. *Faults*: Collector restarts, network errors, remote store problems etc.
+
+Log loss can be avoided for _temporary_ faults or overloads.
+Persistent buffers and/or acknowledgements can store or re-send lost in-memory data.
+
+_Sustained overload_ lasting long enough to exceed buffering capacity _will_ cause data loss.
+The only remedy is to ensure that the log collector and store can keep up with the rate of log production.
+
+#### Acknowledgements
+
+[End-to-end acknowledgement](https://vector.dev/docs/about/under-the-hood/architecture/end-to-end-acknowledgements) 
+means that source acknowledgements are _delayed_ until all relevant sinks have received the data.
+After restart, the source can re-send data that did not reach all sinks - the source acts as a persistent buffer.
+
+This is more efficient than duplicating the data again in a vector disk buffer,
+but only works for sources that support acknowledgement and "at-least-once" reliable delivery.
+
+Examples of acknowledgement sources:
```

Comment:
End-to-end acknowledgement involves both the source and the sink. The source defines how vector acknowledges to the sink that it can "forget" data that has been sucessfully sent. E.g. for HTTP it's a HTTP response, for a file source it is moving the persistent "position". The sink is involved because it receives ack that the data was sucessfully sent. The source delays ack to the remote sink until all relevant sources have indicated that they have sent OK. All sinks support end-to-end ack, even if the sink protocol is not "reliable" - they report when the data is sent "to the best of their knowledge". Only some sources support it,  it requires that the remote source will "re-send" unacked messages. In the case of files this just means vector will re-read the file, for network protcols it's an actual re-send after re-connect. I'll clarify in the doc.